File: Neetcode-150/Stack/CarFleet.py
# Time Complexity: O(nlogn)
# Space Complexity: O(n)
'''
yeah fuck htis problem too. I just did whatever and watched the soln lol, but the carMap and the stack will always 
increase linearly even though I had to use 2 extra data structures(regretably). The time complexity will be O(n)
because we're sorting the position array.
'''

def carFleet(target, position, speed):

    
    carMap = {}

    for i in range(len(position)):
        carMap[position[i]] = speed[i]


    stack = []

    for pos in sorted(position, reverse=True):
        stack.append((target - pos) / carMap[pos])
        if(len(stack) >= 2 and stack[-1] <= stack[-2]):
            stack.pop()
        print(stack)

    return len(stack)

print(carFleet(12, [1,4], [3,2]))

        
        

File: Neetcode-150/Stack/car_fleet.java
import java.util.Arrays;
import java.util.HashMap;

public class car_fleet {
    public static int carFleet(int target, int[] position, int[] speed) {
        int fleets = 0;
        double time = 0, dest_time = 0; // to store the current time and the destination time

        HashMap<Integer, Integer> map = new HashMap<>(); // use a hashmap to store the position and its speed

        for (int i = 0; i < position.length; i++) { // add to hashmap
            map.put(position[i], speed[i]);
        }

        Arrays.sort(position); // sort the position array so know what cars are ahead and which is last

        for (int i = position.length - 1; i >= 0; i--) { // go through the array in reverse because we start from the car closest to target
            // we are using double because integer rounds decimals down but we need the exact value
            dest_time = ((double) target - position[i]) / map.get(position[i]); // calculate destination time using formula: (target - position) / speed
            if (time < dest_time) { // if the current time is less than the destination time
                fleets++; // increment to indicate new fleet
                time = dest_time; // update current time
            }
        }
        
        return fleets;        
    }
    public static void main(String[] args) {
        int[] position = {1,4}, speed = {3,2}, position2 = {4,1,0,7}, speed2 = {2,2,1,1};
        int target = 10;

        System.out.println(carFleet(target, position, speed));
        System.out.println(carFleet(target, position2, speed2));

        // time complexity will be O(nlogn) since we are sorting the positions array and the space complexity will be O(n) for the hashmap which will be of n elements
    }
}